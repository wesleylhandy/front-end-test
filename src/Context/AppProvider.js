import React, { Component } from "react";
import { callApi } from "../Utils/fetch";

export const AppContext = React.createContext();

const ADD_NOTES = "ADD_NOTES";
const SELECT_NOTE = "SELECT_NOTE";
const MARK_AS_READ = "MARK_AS_READ";

const reducer = (state, action) => {
	const { type, notes, currentNoteIndex } = action;
	switch (type) {
		case ADD_NOTES:
			return {
				...state,
				notes,
			};
		case SELECT_NOTE:
			return {
				...state,
				currentNoteIndex,
			};
		case MARK_AS_READ:
			const notesState = [...state.notes];
			notesState[state.currentNoteIndex].read = true;
			return {
				...state,
				notes: notesState,
			};
		default:
			return { ...state };
	}
};

class AppProvider extends Component {
	state = {
		notes: [],
		currentNoteIndex: -1,
		getNotes: async () => {
			try {
                const notes = await callApi("/notes", { method: "GET" });
				this.setState(state => reducer(state, { type: ADD_NOTES, notes }));
			} catch (e) {
				console.error({ e });
				// show alert???
			}
		},
		selectNote: idx => {
			this.setState(state =>
				reducer(state, { type: SELECT_NOTE, currentNoteIndex: idx })
			);
		},
		markAsRead: () => {
			// this function is called by NoteDetails, which displays the currentNoteIndex
			// so this should only mark as read the current note index
			this.setState(state => reducer(state, { type: MARK_AS_READ }));
		},
	};
	render() {
		const {
			state,
			props: { children },
		} = this;
		return <AppContext.Provider value={state}>{children}</AppContext.Provider>;
	}
}

export default AppProvider;
