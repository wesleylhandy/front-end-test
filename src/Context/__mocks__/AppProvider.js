import React, { Component } from "react";
const fetchMock = require("fetch-mock");

import { callApi } from "../../Utils/fetch";
import testBody from "../../../db.json";

export const AppContext = React.createContext();

const ADD_NOTES = "ADD_NOTES";
const SELECT_NOTE = "SELECT_NOTE";
const TOGGLE_READ_STATUS = "TOGGLE_READ_STATUS";

const reducer = (state, action) => {
	const { type, notes, currentNoteIndex } = action;
	switch (type) {
		case ADD_NOTES:
			return {
				...state,
				notes,
			};
		case SELECT_NOTE:
			return {
				...state,
				currentNoteIndex,
			};
		case TOGGLE_READ_STATUS:
			const notesState = [...state.notes];
			notesState[state.currentNoteIndex].read = !notesState[
				state.currentNoteIndex
			].read;
			return {
				...state,
				notes: notesState,
			};
		default:
			return { ...state };
	}
};

class AppProvider extends Component {
	state = {
		notes: [],
		currentNoteIndex: -1,
		getNotes: async () => {
			try {
				fetchMock.get("http://fake.com", testBody.notes);
                const notes = await callApi("http://fake.com", { method: "GET" });
				this.setState(state => reducer(state, { type: ADD_NOTES, notes }));
			} catch (e) {
				console.error({ e });
				// show alert???
			}
		},
		selectNote: idx => {
			this.setState(state =>
				reducer(state, { type: SELECT_NOTE, currentNoteIndex: idx })
			);
		},
		toggleReadStatus: () => {
			// this function is called by NoteDetails, which displays the currentNoteIndex
			// so this should only mark as read the current note index
			this.setState(state => reducer(state, { type: TOGGLE_READ_STATUS }));
		},
	};
	render() {
		const {
			state,
			props: { children },
		} = this;
		return <AppContext.Provider value={state}>{children}</AppContext.Provider>;
	}
}

export default AppProvider;
