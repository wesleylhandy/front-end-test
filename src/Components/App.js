import React, { Component } from "react";
import { AppContext } from "../Context/AppProvider";
import Header from "./Header";
import Sidebar from "./Sidebar";
import NoteDetails from "./NoteDetails";

class App extends Component {
	static contextType = AppContext;

	componentWillMount() {
		this.context.getNotes();
	}

	render() {
		const { notes, markAsRead, selectNote, currentNoteIndex } = this.context;
		const currentNote =
			notes.length && currentNoteIndex > -1 ? notes[currentNoteIndex] : {};
		const unreadCount = notes.reduce((count, {read}) => read ? count : count + 1, 0);
		return (
			<div className="App">
				<Header unreadCount={unreadCount} />
				<Sidebar
					notes={notes}
					selectNote={selectNote}
					currentNoteIndex={currentNoteIndex}
				/>
				<NoteDetails note={currentNote} markAsRead={markAsRead} />
			</div>
		);
	}
}

export default App;
