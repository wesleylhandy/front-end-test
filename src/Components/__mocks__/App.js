import React, { Component } from "react";
import { AppContext } from "../../Context/__mocks__/AppProvider";
import Header from "../Header";
import Sidebar from "../Sidebar";
import NoteDetails from "../NoteDetails";
import ErrorBoundary from "../ErrorBoundary";

import "../app.css";

class App extends Component {
	static contextType = AppContext;

	componentDidMount() {
		this.context.getNotes();
	}

	render() {
		const {
			notes,
			toggleReadStatus,
			selectNote,
			currentNoteIndex,
		} = this.context;
		const currentNote =
			notes.length && currentNoteIndex > -1 ? notes[currentNoteIndex] : {};
		const unreadCount = notes.reduce(
			(count, { read }) => (read ? count : count + 1),
			0
		);
		return (
			<ErrorBoundary>
				<div className="App">
					<Header unreadCount={unreadCount} />
					<Sidebar
						notes={notes}
						selectNote={selectNote}
						currentNoteIndex={currentNoteIndex}
					/>
					<NoteDetails note={currentNote} toggleReadStatus={toggleReadStatus} />
				</div>
			</ErrorBoundary>
		);
	}
}

export default App;
