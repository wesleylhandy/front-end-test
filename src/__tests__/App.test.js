import React from "react";
import { create } from "react-test-renderer";
import App from "../Components/__mocks__/App";
import AppProvider from "../Context/__mocks__/AppProvider";

/**** NEEDS IMPROVEMENT ****/
/*	This tests the functionality of two mock component
/*  that match the current state of the components
/*  but won't catch changes...
/*  I'm not all that certain how to test Context Api properly
/**** 					****/

it("renders correctly", () => {
	const tree = create(
		<AppProvider>
			<App />
		</AppProvider>
	).toJSON();
	expect(tree).toMatchSnapshot();
});
