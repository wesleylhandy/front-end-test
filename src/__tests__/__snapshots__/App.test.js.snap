// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <header
    className="header"
  >
    <img
      alt="logo"
      className="header__logo"
      src="logo.svg"
    />
    <h1
      className="header__title"
    >
      Notes Viewer Test App
    </h1>
    <div
      className="header__unread-count"
    >
      Unread:
      <span
        className="header__unread-count--count"
      >
        0
      </span>
    </div>
  </header>
  <section
    className="NotesSidebar"
  >
    <h2
      className="NotesSidebar__title"
    >
      Available Notes:
    </h2>
    <div
      className="NotesSidebar__list"
    >
      <div
        className="NotesSidebarItem empty"
      >
        --Empty--
      </div>
    </div>
  </section>
</div>
`;
